### PR TITLE
SONAR-7723 WS api/permissions/add_project_creator_to_template

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/permission/ws/template/AddProjectCreatorToTemplateAction.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+  private void updateTemplatePermission(DbSession dbSession, PermissionTemplatePermissionDto templatePermission) {
+    PermissionTemplatePermissionDto targetTemplatePermission = templatePermission
+      .setUpdatedAt(system.now())
+      .setWithProjectCreator(true);
+    dbClient.permissionTemplatePermissionDao().update(dbSession, targetTemplatePermission);
+    dbSession.commit();
+  }
+
+  private AddProjectCreatorToTemplateWsRequest toWsRequest(Request request) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/ws/template/AddProjectCreatorToTemplateAction.java]: MAJOR Make "toWsRequest" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/permission/ws/template/AddProjectCreatorToTemplateAction.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    checkGlobalAdminUser(userSession);
+    doHandle(toWsRequest(request));
+    response.noContent();
+  }
+
+  private void doHandle(AddProjectCreatorToTemplateWsRequest request) {
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      PermissionTemplateDto template = dependenciesFinder.getTemplate(dbSession, WsTemplateRef.newTemplateRef(request.getTemplateId(), request.getTemplateName()));
+      Optional<PermissionTemplatePermissionDto> templatePermission = dbClient.permissionTemplatePermissionDao().selectByPermissionAndTemplateId(dbSession, request.getPermission(), template.getId());
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/ws/template/AddProjectCreatorToTemplateAction.java]: MINOR Split this 198 characters long line (which is greater than 180 authorized). [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

File: sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplatePermissionDao.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    return dto;
+  }
+
+  public PermissionTemplatePermissionDto update(DbSession dbSession, PermissionTemplatePermissionDto templatePermissionDto) {
+    requireNonNull(templatePermissionDto.getId());
+    mapper(dbSession).update(templatePermissionDto);
+    return templatePermissionDto;
+  }
+
+  private PermissionTemplatePermissionMapper mapper(DbSession dbSession) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplatePermissionDao.java]: MAJOR Make "mapper" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* MAJOR 3 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR org.sonarsource.sonarqube:sonar-db:src/main/java/org/sonar/db/permission/template: Add a 'package-info.java' file to document the 'template' package [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

File: server/sonar-web/src/main/webapp/WEB-INF/db/migrate/1206_create_perm_templates_permissions.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+# along with this program; if not, write to the Free Software Foundation,
+# Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+#
+# SonarQube 6.0
+#
+class CreatePermTemplatesPermissions < ActiveRecord::Migration
+
+  def self.up
+    create_table 'perm_tpl_permissions' do |t|
```
Comment (julienlancelot) [inline server/sonar-web/src/main/webapp/WEB-INF/db/migrate/1206_create_perm_templates_permissions.rb]: Have you validate the structure with the Atlas team

Comment (julienlancelot) [inline server/sonar-web/src/main/webapp/WEB-INF/db/migrate/1206_create_perm_templates_permissions.rb]: Hum, the name is strange, isn't it ? Permission template permissions ?

Comment (julienlancelot) [inline server/sonar-web/src/main/webapp/WEB-INF/db/migrate/1206_create_perm_templates_permissions.rb]: It should be 'perm_tpl_characteristic'

File: sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplatePermissionDto.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.db.permission.template;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class PermissionTemplatePermissionDto {
+
+  private static final int MAX_PERMISSION_KEY_LENGTH = 64;
```
Comment (julienlancelot) [inline sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplatePermissionDto.java]: This constant would probably need to be placed elsewhere, but for the moment it's ok here.

File: sonar-db/src/main/resources/org/sonar/db/permission/template/PermissionTemplatePermissionMapper.xml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    ptp.with_project_creator as withProjectCreator,
+    ptp.created_at as createdAt,
+    ptp.updated_at as updatedAt
+  </sql>
+
+  <select id="selectByTemplateId" parameterType="long" resultType="TemplatePermission">
+    select
+    <include refid="columns" />
+    from perm_tpl_permissions ptp
+    where ptp.template_id=#{templateId}
```
Comment (julienlancelot) [inline sonar-db/src/main/resources/org/sonar/db/permission/template/PermissionTemplatePermissionMapper.xml]: When validating the schema with Atlas, tell them that you need to query this table by template_id only (as there's no index on it)

**Comment by julienlancelot** [comment]
Please update the table name, send an email to Atlas about table structure and LGTM

**Pull request merged**